logback-classic/src/main/java9/module-info.java

```diff
@@ -1,4 +1,5 @@
 module ch.qos.logback.classic { 
+  requires java.logging;
```

Comment:
This forces the module requirement for everybody using Logback Classic in a modular environment. This statement should be requires static instead.

Reply:
@iajn Please help me understand your concern.

1. `LevelChangePropagator` is part of the logback classic JAR file. It isn't an optional plugin like the JUL to SLF4J bridge.
2. `requires` means that any code that depends on logback classic will have to provide `java.logging` on the modulepath, which coincidentally doesn't impact the end-user at all since this is guaranteed to ship with the JDK.
3. While it is true that not everyone who uses logback classic will choose to use `LevelChangePropagator` if you decide to use `requires static` every time this happens you will quickly find yourself requiring nothing. As far as I understand it, this isn't what `requires static` was meant for.
4. And finally, `requires` is not `requires transitive` so we're not saying that anyone who depends on logback classic will propagate `requires java.logging` to its own users.

Am I missing anything?

Reply:
I was mostly concerned for situations where the end user uses jlink to shave off modules that are not required to slimline the JDK delivered with the application. When used with a full JDK that always includes the java.logging module I don't think it matters whether requires or requires static is used.

I'm not entirely sure what you mean with "requiring nothing", but to [quote this blog post](https://nipafx.dev/java-modules-optional-dependencies/):
_The module system allows optional dependencies with the requires static clause. Module required this way are accessible at compile time but can be absent at run time._

To me this sounds like a good use case. If the end user doesn't use the LevelChangePropagator, they also don't require the java.logging module, so it can be absent from the jlink'd JDK. If the LevelChangePropagator is used, the user needs to add the requirement for java.logging themselves.

Because Logback is usually configured in xml instead of code, this does add a minor annoyance of needing to add java.logging to the module-info (or to the --add-modules list) manually without there being any indication in the code to do so. The application should work fine on a full JDK without the explicit requirement, but would probably encounter a runtime error when using a jlink'd JDK. This requirement could of course be explicitly stated in the documentation that covers the LevelChangePropagator.

There are minor cons for both approaches. In my opinion it would be better to not force the extra module (even though it is very small) for everybody. But this is not a hill I wish to die on.

Reply:
Thank you both for your thoughtful comments which are highly appreciated. I think the Jlink argument brings a new perspective which many of us  previously did not pay attention to.